cmd/thv-operator/controllers/virtualmcpserver_controller.go

```diff
@@ -896,6 +906,169 @@ func createVmcpServiceURL(vmcpName, namespace string, port int32) string {
+// For ConfigMap mode (inline), secrets are referenced as environment variables that will be
+// mounted in the deployment.
+func (r *VirtualMCPServerReconciler) convertExternalAuthConfigToStrategy(
+	ctx context.Context,
+	namespace string,
+	externalAuthConfig *mcpv1alpha1.MCPExternalAuthConfig,
+) (*authtypes.BackendAuthStrategy, error) {
+	// Use the converter registry to convert to typed strategy
+	registry := converters.DefaultRegistry()
+	converter, err := registry.GetConverter(externalAuthConfig.Spec.Type)
+	if err != nil {
+		return nil, err
+	}
+
+	// Convert to typed BackendAuthStrategy (this will use env var references for secrets)
+	strategy, err := converter.ConvertToStrategy(externalAuthConfig)
+	if err != nil {
+		return nil, fmt.Errorf("failed to convert external auth config to strategy: %w", err)
+	}
+
+	// For header injection, resolve secrets from Kubernetes
+	if externalAuthConfig.Spec.Type == mcpv1alpha1.ExternalAuthTypeHeaderInjection {
+		strategy, err = converter.ResolveSecrets(ctx, externalAuthConfig, r.Client, namespace, strategy)
+		if err != nil {
+			return nil, fmt.Errorf("failed to resolve secrets for header injection: %w", err)
+		}
+	}
+
+	return strategy, nil
+}
+
+// convertBackendAuthConfigToVMCP converts a BackendAuthConfig from CRD to vmcp config.
+func (r *VirtualMCPServerReconciler) convertBackendAuthConfigToVMCP(
+	ctx context.Context,
+	namespace string,
+	crdConfig *mcpv1alpha1.BackendAuthConfig,
+) (*authtypes.BackendAuthStrategy, error) {
+	// For type="discovered", return a minimal strategy (will be populated by discovery)
+	if crdConfig.Type == mcpv1alpha1.BackendAuthTypeDiscovered {
+		return &authtypes.BackendAuthStrategy{
+			Type: crdConfig.Type,
+		}, nil
+	}
+
+	// For type="external_auth_config_ref", fetch and convert the referenced config
+	if crdConfig.ExternalAuthConfigRef != nil {
+		// Fetch the MCPExternalAuthConfig and convert it
+		externalAuthConfig, err := ctrlutil.GetExternalAuthConfigByName(
+			ctx, r.Client, namespace, crdConfig.ExternalAuthConfigRef.Name)
+		if err != nil {
+			return nil, fmt.Errorf("failed to get MCPExternalAuthConfig %s: %w", crdConfig.ExternalAuthConfigRef.Name, err)
+		}
+
+		// Convert the external auth config to strategy
+		return r.convertExternalAuthConfigToStrategy(ctx, namespace, externalAuthConfig)
+	}
+
+	// Fallback: return minimal strategy
+	return &authtypes.BackendAuthStrategy{
+		Type: crdConfig.Type,
+	}, nil
+}
+
+// discovers ExternalAuthConfig from MCPServers and adds them to the outgoing config
+func (r *VirtualMCPServerReconciler) discoverExternalAuthConfigs(
+	ctx context.Context,
+	vmcp *mcpv1alpha1.VirtualMCPServer,
+	workloadNames []string,
+	outgoing *vmcpconfig.OutgoingAuthConfig,
+) {
+	ctxLogger := log.FromContext(ctx)
+
+	for _, workloadName := range workloadNames {
+		mcpServer := &mcpv1alpha1.MCPServer{}
+		if err := r.Get(ctx, types.NamespacedName{Name: workloadName, Namespace: vmcp.Namespace}, mcpServer); err != nil {
```

Comment:
I wonder if it would be easier to do a full `List` and then filter by names rather than doing N Get calls for the names. But that's an optimization for layer maybe..worth a comment with a TODO now?